src/v/transform/logging/log_manager.cc

```diff
@@ -0,0 +1,294 @@
+                    it->second.emplace_back(
+                      model::transform_name{std::move(name)},
+                      std::move(ev_json));
+                    return ss::now();
+                });
+          });
+
+        co_await ss::max_concurrent_for_each(
+          batches, MAX_CONCURRENCY, [this](auto& pr) mutable -> ss::future<> {
+              auto& [pid, evs] = pr;
+              return do_flush(pid, std::move(evs));
+          });
+
+        vlog(tlg_log.trace, "Processed {} log events", n_events);
+        co_return;
+    }
+
+    ss::future<ss::chunked_fifo<iobuf>>
+    do_serialize_log_events(model::transform_name_view name, auto events) {
+        ss::chunked_fifo<iobuf> ev_json;
+        ev_json.reserve(events.size());
+
+        while (!events.empty()) {
+            auto e = std::move(events.front());
+            events.pop_front();
+            iobuf b;
+            e.event.to_json(name, b);
+            ev_json.emplace_back(std::move(b));
+
+            // reactor will stall if we try to serialize a whole lot of
+            // messages all at once
+            co_await ss::maybe_yield();
+        }
+        // semaphore units for the primary buffer are free at this point.
+
+        co_return std::move(ev_json);
+    }
+
+    ss::future<> do_flush(model::partition_id pid, io::json_batches events) {
+        // TODO(oren): it might be a good idea to cap the amount of serialized
+        // log data in flight at one time.
+        // maybe batching actually happens here
+
+        co_await _client->write(pid, std::move(events));
+        co_return;
+    }
+
+private:
+    ss::abort_source* _as = nullptr;
+    transform::logging::client* _client = nullptr;
+    config::binding<std::chrono::milliseconds> _interval_ms;
+    simple_time_jitter<ClockType> _jitter;
+    ss::condition_variable _wakeup_signal{};
+};
+
+template class flusher<ss::lowres_clock>;
+template class flusher<ss::manual_clock>;
+
+} // namespace detail
+
+template<typename ClockType>
+manager<ClockType>::manager(
+  model::node_id self,
+  std::unique_ptr<client> c,
+  size_t bc,
+  config::binding<size_t> ll,
+  config::binding<std::chrono::milliseconds> fi)
+  : _self(self)
+  , _client(std::move(c))
+  , _line_limit_bytes(std::move(ll))
+  , _buffer_limit_bytes(bc)
+  , _buffer_low_water_mark(_buffer_limit_bytes / lwm_denom)
+  , _buffer_sem(_buffer_limit_bytes, "Log manager buffer semaphore")
+  , _flush_interval_ms(std::move(fi)) {}
+
+template<typename ClockType>
+manager<ClockType>::~manager() = default;
+
+template<typename ClockType>
+ss::future<> manager<ClockType>::start() {
+    _flusher = std::make_unique<detail::flusher<ClockType>>(
+      &_as, _client.get(), _flush_interval_ms);
+    return _flusher->template start<>(_gate, &_log_buffers);
+}
+
+template<typename ClockType>
+ss::future<> manager<ClockType>::stop() {
+    _as.request_abort();
+    co_await _flusher->stop();
```

Comment:
Should we be closing the gate first?

Reply:
I think we need to break the semaphore first because the flush fiber is holding the gate open, right?

incidentally, the fact that this looks strange is probably a bit of an API smell for the flusher thing...

Reply:
Why does this class even need the gate?

Reply:
I think the suggestion is "push the gate down into `flusher`", which sounds right to me, I've done that.

Reply:
It was a question, but a reasonable answer is that it doesn't and it can be pushed into the flusher :)